apps/extension/src/storage/base.ts

```diff
@@ -20,30 +20,55 @@ export interface StorageItem<T> {
 }
 
 type Version = string;
+
+export type MigrationMap<OldState, NewState> = {
+  [K in keyof OldState & keyof NewState]?: (
+    prev: OldState[K],
+    get: <K extends keyof NewState>(key: K) => Promise<NewState[K]>,
+  ) => NewState[K] | Promise<NewState[K]>;
+};
+
 // It is quite difficult writing a generic that covers all migration function kinds.
 // Therefore, the writer of the migration should ensure it is typesafe when they define it.
-// See `migration.test.ts` for an example.
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
-type Migration = Record<Version, (a: any) => any>;
-type Migrations<K extends string | number | symbol> = Partial<Record<K, Migration>>;
+export type Migrations<T> = Partial<Record<Version, MigrationMap<any, T>>>;
+
+export type VersionSteps = Record<Version, Version>;
 
 export class ExtensionStorage<T> {
+  private migrationLocks: Record<string, Promise<T[keyof T]>> = {};
```

Comment:
thinking more about migration locks, isn't there a risk of blocking if a migration gets stuck or slow? perhaps we should add a timeout mechanism to the migration locks

Reply:
**update:** the comment should be updated to reflect that the synchronization mechanism prevents concurrent modifications to the _entire_ storage object, not just a single field.

In terms of the risk of blocking, there now might be a risk of blocking between _storage versions_ if the migration gets stuck or slow for a specific storage object (rather than a partial storage key).

Reply:
Talked offline. Not relevant to our specific migration, but think timeouts in the future could be a good feature to add.